File: python_concept/embed_gen.py
#!/usr/bin/env python3
"""embed_gen.py — corpus .txt -> base de vetores-histograma em JSON.

Materializa a INDEXACAO offline: fatia o texto em chunks, tokeniza cada um contra o
vocabulario de tokens (ordem FIXA) e grava o histograma esparso (bag-de-tokens = o
"embed" do chunk). E a base de busca, persistida e reproduzivel.

Depende da lib **sylkit** (na raiz, ao lado) — o tokenizer/vetorizacao ja organizados.
O JSON e' AUTO-SUFICIENTE: carrega o vocabulario (vocab) inline e o idf global, entao
da' pra reconstruir o tf-idf e o cosseno sem reprocessar o corpus. Campos em ingles.

Layout:
  meta      -> corpus, chunk_size, contagens, coverage, generator, tokens_file, vocab[]
  idf       -> {dim: log(N/df)} idf global por dimensao vista
  chunks[]  -> {id, start (byte), len, tokens, oov, norm, text?, vec {dim: count}}

Uso:
    python3 embed_gen.py sda.txt                       # chunk 2048, sda-tokenized.json
    python3 embed_gen.py sda.txt --chunk 1024 --saida sda-1k.json
    python3 embed_gen.py sda.txt --sem-texto --max-chunks 50
"""
import os, sys, json, time, argparse, glob

# sylkit ao lado deste arquivo (em python_concept/) — import direto
from sylkit import load_vocab, load_driver, histogram, tfidf, compute_idf, chunk_text

# raiz do projeto = um nível acima; pastas por papel (independe do cwd)
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DRIVERS = os.path.join(ROOT, "drivers")          # vocabularios de tokens
IMPORTABLES = os.path.join(ROOT, "importables")  # corpora crus pra ingerir
RAGFILES = os.path.join(ROOT, "ragfiles")        # bases RAG geradas


def list_drivers(drivers_dir):
    """Lista os drivers .drv instalados (nome, idioma, nº silabas, nº keywords)."""
    out = []
    for path in sorted(glob.glob(os.path.join(drivers_dir, "*.drv"))):
        vocab, keywords = load_driver(path)
        fname = os.path.basename(path)
        # tokens_<Lang>_PTBR.drv -> idioma = <Lang>
        lang = fname[len("tokens_"):-len(".drv")] if fname.startswith("tokens_") else fname[:-4]
        lang = lang[:-len("_PTBR")] if lang.endswith("_PTBR") else lang
        header = ""
        with open(path, encoding="utf-8") as f:
            first = f.readline().strip()
            header = first[1:].strip() if first.startswith("#") else ""
        out.append({
            "name": fname,
            "language": lang,
            "syllables": len(vocab) - len(keywords),
            "keywords": len(keywords),
            "vocab_size": len(vocab),
            "header": header,
        })
    return {"drivers_dir": drivers_dir, "count": len(out), "drivers": out}


def main():
    ap = argparse.ArgumentParser(
        description="Gera base de vetores-histograma (JSON) de um corpus.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="exemplos (defaults: corpus em importables/, tokens em drivers/, saida em ragfiles/):\n"
               "  python3 embed_gen.py importables/sda.txt           # -> ragfiles/sda-tokenized.json\n"
               "  python3 embed_gen.py importables/outro.txt --chunk 1024\n"
               "  python3 embed_gen.py --sem-texto                   # usa o corpus default (importables/sda.txt)\n")
    ap.add_argument("corpus", nargs="?", default=os.path.join(IMPORTABLES, "sda.txt"),
                    help="corpus .txt de entrada (default importables/sda.txt)")
    ap.add_argument("--chunk", type=int, default=2048, help="bytes por chunk (default 2048)")
    ap.add_argument("--saida", default=None, help="JSON de saida (default ragfiles/{corpus}-tokenized.json)")
    ap.add_argument("--tokens", default=os.path.join(DRIVERS, "tokens_PTBR.drv"),
                    help="driver de tokens .drv (default drivers/tokens_PTBR.drv)")
    ap.add_argument("--list-drivers", action="store_true",
                    help="lista os drivers .drv instalados em drivers/ (saida JSON) e sai")
    ap.add_argument("--max-chunks", type=int, default=0, help="limita chunks (0 = corpus inteiro)")
    ap.add_argument("--sem-texto", action="store_true", help="nao grava o texto de cada chunk")
    ap.add_argument("--compacto", action="store_true",
                    help="grava minificado (1 linha); default e' indentado com TAB p/ ler no editor")
    ap.add_argument("--quiet", "-q", action="store_true",
                    help="suprime output de status (so gera o JSON)")
    if len(sys.argv) == 1:        # sem argumentos -> mostra o help (nao dispara a geracao)
        ap.print_help()
        sys.exit(0)
    args = ap.parse_args()

    if args.list_drivers:
        print(json.dumps(list_drivers(DRIVERS), ensure_ascii=False, indent=2))
        sys.exit(0)

    stem = os.path.splitext(os.path.basename(args.corpus))[0]
    saida = args.saida or os.path.join(RAGFILES, f"{stem}-tokenized.json")
    vocab, idx = load_vocab(args.tokens)
    with open(args.corpus, encoding="utf-8") as f:
        text = f.read()

    t0 = time.time()
    # fatia preservando o byte de inicio de cada chunk (pra rastrear no texto original)
    pieces = chunk_text(text, args.chunk)
    if args.max_chunks:
        pieces = pieces[:args.max_chunks]

    # passada 1: histograma (tf) + cobertura de cada chunk
    chunks, tfs = [], []
    tot_tokens = tot_oov = 0
    cursor = 0
    for cid, piece in enumerate(pieces):
        start = text.find(piece, cursor)          # offset real no texto original
        if start < 0:
            start = cursor
        cursor = start + len(piece)
        tf, stats = histogram(piece, idx, with_stats=True)
        tfs.append(tf)
        tot_tokens += stats["total"]; tot_oov += stats["oov"]
        chunks.append({"id": cid, "start": start, "len": len(piece),
                       "tokens": stats["total"], "oov": stats["oov"], "vec": tf,
                       "text": None if args.sem_texto else piece})

    # idf global = log(N/df): token em todo chunk -> idf~0 (vira stopword)
    idf = compute_idf(tfs, len(pieces))

    # passada 2: norma tf-idf de cada chunk (pra cosseno rapido depois)
    for c in chunks:
        _, norm = tfidf(c["vec"], idf)
        c["norm"] = round(norm, 6)
        # JSON exige chaves string; ordena por dim p/ legibilidade/diff estavel
        c["vec"] = {str(d): n for d, n in sorted(c["vec"].items())}

    base = {
        "meta": {
            "corpus": os.path.basename(args.corpus),
            "bytes": len(text.encode("utf-8")),
            "chunk_size": args.chunk,
            "n_chunks": len(pieces),
            "vocab_size": len(vocab),
            "vocab_used": len(idf),
            "tokens_total": tot_tokens,
            "oov_total": tot_oov,
            "coverage": round(1 - tot_oov / tot_tokens, 4) if tot_tokens else 0.0,
            "with_text": not args.sem_texto,
            "generator": os.path.basename(__file__),       # quem gerou esta base
            "tokens_file": os.path.basename(args.tokens),  # qual vocabulario foi usado
            "built_at": time.strftime("%Y-%m-%dT%H:%M:%S"),
            "vocab": vocab,                                 # a matriz inline (ordem imutavel)
        },
        "idf": {str(d): round(v, 6) for d, v in sorted(idf.items())},
        "chunks": chunks,
    }

    with open(saida, "w", encoding="utf-8") as f:
        if args.compacto:
            json.dump(base, f, ensure_ascii=False)
        else:
            json.dump(base, f, ensure_ascii=False, indent="\t")  # TAB: visivel no vim/vscode
    dt = time.time() - t0
    size = os.path.getsize(saida)

    if not args.quiet:
        print(f"== EMBEDDINGS GERADOS ==")
        print(f"corpus:    {args.corpus}  ({base['meta']['bytes']:,} bytes)")
        print(f"chunks:    {base['meta']['n_chunks']}  (~{args.chunk} B cada)  em {dt:.2f}s")
        print(f"tokens:    {base['meta']['tokens_file']}  ->  {base['meta']['vocab_size']} dims "
              f"({base['meta']['vocab_used']} usadas)")
        print(f"silabas:   {tot_tokens:,}  (OOV {tot_oov:,}  ->  cobertura {base['meta']['coverage']*100:.2f}%)")
        print(f"saida:     {saida}  ({size:,} bytes, {'com' if not args.sem_texto else 'sem'} texto)")


if __name__ == "__main__":
    main()

File: rust_concept/src/bin/embed_gen.rs
//! embed_gen (Rust) — corpus .txt -> base de vetores-histograma em JSON.
//! Porte do embed_gen.py: mesmo schema (campos EN, vec esparso, idf, vocab inline).
use std::collections::HashMap;
use std::time::Instant;
use serde_json::{Map, Value};
use sylkit::{chunk_text, compute_idf, find_chars, histogram, load_vocab, tfidf_norm};

fn help() {
    println!(
"embed_gen (Rust) — gera base de vetores-histograma (JSON) de um corpus.

uso:
  embed_gen <corpus.txt> [opcoes]

opcoes:
  --chunk N        bytes(chars) por chunk (default 2048)
  --saida ARQ      JSON de saida (default {{corpus}}-tokenized.json)
  --tokens ARQ     vocabulario de tokens (default tokens_PTBR.txt)
  --max-chunks N   limita chunks (0 = corpus inteiro)
  --sem-texto      nao grava o texto de cada chunk
  --compacto       JSON minificado (default indentado com TAB)

exemplos:
  embed_gen sda.txt
  embed_gen sda.txt --chunk 1024 --saida sda-1k.json");
}

fn round_to(x: f64, d: i32) -> f64 {
    let f = 10f64.powi(d);
    (x * f).round() / f
}

fn commas(n: usize) -> String {
    let s = n.to_string();
    let len = s.len();
    let mut out = String::new();
    for (i, ch) in s.chars().enumerate() {
        if i > 0 && (len - i) % 3 == 0 { out.push(','); }
        out.push(ch);
    }
    out
}

fn iso_utc_now() -> String {
    let secs = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH).unwrap().as_secs() as i64;
    let days = secs.div_euclid(86400);
    let rem = secs.rem_euclid(86400);
    let (h, mi, s) = (rem / 3600, (rem % 3600) / 60, rem % 60);
    let z = days + 719468;
    let era = (if z >= 0 { z } else { z - 146096 }) / 146097;
    let doe = z - era * 146097;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let mut y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let m = if mp < 10 { mp + 3 } else { mp - 9 };
    if m <= 2 { y += 1; }
    format!("{:04}-{:02}-{:02}T{:02}:{:02}:{:02}", y, m, d, h, mi, s)
}

fn basename(p: &str) -> String { p.rsplit('/').next().unwrap_or(p).to_string() }

struct Rec { id: usize, start: usize, len: usize, tokens: usize, oov: usize,
             tf: HashMap<usize, u32>, text: Option<String> }

fn main() {
    let argv: Vec<String> = std::env::args().collect();
    if argv.len() == 1 { help(); return; }

    let (mut corpus, mut chunk, mut saida) = ("sda.txt".to_string(), 2048usize, None::<String>);
    let (mut tokens, mut max_chunks) = ("tokens_PTBR.txt".to_string(), 0usize);
    let (mut sem_texto, mut compacto) = (false, false);
    let mut pos: Vec<String> = vec![];
    let mut it = argv[1..].iter();
    while let Some(a) = it.next() {
        match a.as_str() {
            "-h" | "--help" => { help(); return; }
            "--chunk" => chunk = it.next().unwrap().parse().expect("--chunk N"),
            "--saida" | "--out" => saida = Some(it.next().unwrap().clone()),
            "--tokens" => tokens = it.next().unwrap().clone(),
            "--max-chunks" => max_chunks = it.next().unwrap().parse().expect("--max-chunks N"),
            "--sem-texto" => sem_texto = true,
            "--compacto" => compacto = true,
            other => pos.push(other.to_string()),
        }
    }
    if let Some(c) = pos.into_iter().next() { corpus = c; }
    let saida = saida.unwrap_or_else(|| {
        let b = basename(&corpus);
        let stem = b.rsplit_once('.').map(|(s, _)| s).unwrap_or(&b);
        format!("{}-tokenized.json", stem)
    });

    let (vocab, index) = load_vocab(&tokens).unwrap_or_else(|e| {
        eprintln!("erro lendo tokens {tokens:?}: {e}"); std::process::exit(1);
    });
    let text = std::fs::read_to_string(&corpus).unwrap_or_else(|e| {
        eprintln!("erro lendo corpus {corpus:?}: {e}"); std::process::exit(1);
    });

    let t0 = Instant::now();
    let chars: Vec<char> = text.chars().collect();
    let mut pieces = chunk_text(&chars, chunk);
    if max_chunks > 0 { pieces.truncate(max_chunks); }

    // passada 1: histograma + offset (char) de cada chunk
    let mut recs: Vec<Rec> = Vec::with_capacity(pieces.len());
    let mut tfs: Vec<HashMap<usize, u32>> = Vec::with_capacity(pieces.len());
    let (mut tot_tokens, mut tot_oov) = (0usize, 0usize);
    let mut cursor = 0usize;
    for (cid, piece) in pieces.iter().enumerate() {
        let needle: Vec<char> = piece.chars().collect();
        let start = find_chars(&chars, &needle, cursor).unwrap_or(cursor);
        cursor = start + needle.len();
        let (tf, total, oov) = histogram(piece, &index);
        tot_tokens += total; tot_oov += oov;
        tfs.push(tf.clone());
        recs.push(Rec { id: cid, start, len: needle.len(), tokens: total, oov,
                        tf, text: if sem_texto { None } else { Some(piece.clone()) } });
    }

    let idf = compute_idf(&tfs, pieces.len());

    // ---- monta o JSON (ordem de campos = embed_gen.py; vec/idf por dim numerico) ----
    let mut meta = Map::new();
    meta.insert("corpus".into(), Value::from(basename(&corpus)));
    meta.insert("bytes".into(), Value::from(text.len()));
    meta.insert("chunk_size".into(), Value::from(chunk));
    meta.insert("n_chunks".into(), Value::from(pieces.len()));
    meta.insert("vocab_size".into(), Value::from(vocab.len()));
    meta.insert("vocab_used".into(), Value::from(idf.len()));
    meta.insert("tokens_total".into(), Value::from(tot_tokens));
    meta.insert("oov_total".into(), Value::from(tot_oov));
    let coverage = if tot_tokens > 0 { round_to(1.0 - tot_oov as f64 / tot_tokens as f64, 4) } else { 0.0 };
    meta.insert("coverage".into(), Value::from(coverage));
    meta.insert("with_text".into(), Value::from(!sem_texto));
    meta.insert("generator".into(), Value::from("embed_gen (rust)"));
    meta.insert("tokens_file".into(), Value::from(basename(&tokens)));
    meta.insert("built_at".into(), Value::from(iso_utc_now()));
    meta.insert("vocab".into(), Value::from(vocab.clone()));

    let mut idf_keys: Vec<usize> = idf.keys().copied().collect();
    idf_keys.sort_unstable();
    let mut idf_map = Map::new();
    for d in &idf_keys {
        idf_map.insert(d.to_string(), Value::from(round_to(idf[d], 6)));
    }

    let chunks: Vec<Value> = recs.iter().map(|r| {
        let mut o = Map::new();
        o.insert("id".into(), Value::from(r.id));
        o.insert("start".into(), Value::from(r.start));
        o.insert("len".into(), Value::from(r.len));
        o.insert("tokens".into(), Value::from(r.tokens));
        o.insert("oov".into(), Value::from(r.oov));
        let mut dims: Vec<usize> = r.tf.keys().copied().collect();
        dims.sort_unstable();
        let mut vec_map = Map::new();
        for d in &dims { vec_map.insert(d.to_string(), Value::from(r.tf[d])); }
        o.insert("vec".into(), Value::Object(vec_map));
        o.insert("text".into(), match &r.text { Some(t) => Value::from(t.clone()), None => Value::Null });
        o.insert("norm".into(), Value::from(round_to(tfidf_norm(&r.tf, &idf), 6)));
        Value::Object(o)
    }).collect();

    let mut root = Map::new();
    root.insert("meta".into(), Value::Object(meta));
    root.insert("idf".into(), Value::Object(idf_map));
    root.insert("chunks".into(), Value::from(chunks));
    let root = Value::Object(root);

    // grava (TAB indent, UTF-8 sem escape ascii — igual ao Python ensure_ascii=False)
    let out = if compacto {
        serde_json::to_string(&root).unwrap()
    } else {
        let mut buf = Vec::new();
        let fmt = serde_json::ser::PrettyFormatter::with_indent(b"\t");
        let mut ser = serde_json::Serializer::with_formatter(&mut buf, fmt);
        use serde::Serialize;
        root.serialize(&mut ser).unwrap();
        String::from_utf8(buf).unwrap()
    };
    std::fs::write(&saida, &out).unwrap();
    let dt = t0.elapsed().as_secs_f64();
    let size = out.len();

    println!("== EMBEDDINGS GERADOS (rust) ==");
    println!("corpus:    {}  ({} bytes)", corpus, commas(text.len()));
    println!("chunks:    {}  (~{} B cada)  em {:.2}s", pieces.len(), chunk, dt);
    println!("tokens:    {}  ->  {} dims ({} usadas)", basename(&tokens), vocab.len(), idf.len());
    println!("silabas:   {}  (OOV {}  ->  cobertura {:.2}%)",
             commas(tot_tokens), commas(tot_oov), coverage * 100.0);
    println!("saida:     {}  ({} bytes, {} texto)",
             saida, commas(size), if sem_texto { "sem" } else { "com" });
}
